// appium-xcuitest-driver/build/lib/commands/alert.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.helpers = exports.commands = void 0;

require("source-map-support/register");

var _appiumBaseDriver = require("appium-base-driver");

var _appiumSupport = require("appium-support");

var _lodash = _interopRequireDefault(require("lodash"));

let commands = {},
    helpers = {},
    extensions = {};
exports.helpers = helpers;
exports.commands = commands;

commands.getAlertText = async function getAlertText() {
  try {
    return await this.proxyCommand('/alert/text', 'GET');
  } catch (err) {
    if (this.isWebContext()) {
      const alert = await this.getAlert();
      return await alert.getText();
    }

    throw err;
  }
};

commands.setAlertText = async function setAlertText(value) {
  if (_lodash.default.isString(value)) {
    value = value.split('');
  }

  try {
    return await this.proxyCommand('/alert/text', 'POST', {
      value
    });
  } catch (err) {
    if (this.isWebContext()) {
      const alert = await this.getAlert();
      return await alert.setText(value);
    }

    throw err;
  }
};

commands.postAcceptAlert = async function postAcceptAlert(opts = {}) {
  try {
    let params = {};

    if (opts.buttonLabel) {
      params.name = opts.buttonLabel;
    }

    return await this.proxyCommand('/alert/accept', 'POST', params);
  } catch (err) {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NoAlertOpenError();
    }

    let alert = await this.getAlert();

    if (alert.close) {
      return await alert.close();
    }

    await alert.ok();
  }
};

commands.postDismissAlert = async function postDismissAlert(opts = {}) {
  try {
    let params = {};

    if (opts.buttonLabel) {
      params.name = opts.buttonLabel;
    }

    return await this.proxyCommand('/alert/dismiss', 'POST', params);
  } catch (err) {
    if (!this.isWebContext()) {
      throw new _appiumBaseDriver.errors.NoAlertOpenError();
    }

    let alert = await this.getAlert();

    if (alert.close) {
      return await alert.close();
    }

    await alert.cancel();
  }
};

commands.getAlertButtons = async function getAlertButtons() {
  try {
    return await this.proxyCommand('/wda/alert/buttons', 'GET');
  } catch (err) {
    throw new _appiumBaseDriver.errors.NoAlertOpenError();
  }
};

commands.mobileHandleAlert = async function mobileHandleAlert(opts = {}) {
  switch (opts.action) {
    case 'accept':
      return await this.postAcceptAlert(opts);

    case 'dismiss':
      return await this.postDismissAlert(opts);

    case 'getButtons':
      return await this.getAlertButtons();

    default:
      throw new Error(`The 'action' value should be either 'accept', 'dismiss' or 'getButtons'. ` + `'${opts.action}' is provided instead.`);
  }
};

helpers.getAlert = async function getAlert() {
  const scrollViews = await this.findNativeElementOrElements('class name', 'XCUIElementTypeScrollView', true);

  if (scrollViews.length !== 1) {
    throw new _appiumBaseDriver.errors.NoAlertOpenError();
  }

  const alert = scrollViews[0];
  const possibleAlertButtons = await this.findNativeElementOrElements('class name', 'XCUIElementTypeButton', true, _appiumSupport.util.unwrapElement(alert));

  if (possibleAlertButtons.length < 1 || possibleAlertButtons.length > 2) {
    throw new _appiumBaseDriver.errors.NoAlertOpenError();
  }

  const assertButtonName = async (button, expectedName = '') => {
    button = _appiumSupport.util.unwrapElement(button);
    const name = await this.proxyCommand(`/element/${button}/attribute/name`, 'GET');

    if ((name === null || name === void 0 ? void 0 : name.toLowerCase()) !== expectedName.toLowerCase()) {
      throw new _appiumBaseDriver.errors.NoAlertOpenError();
    }
  };

  if (possibleAlertButtons.length === 1) {
    const closeButton = possibleAlertButtons[0];
    await assertButtonName(closeButton, 'close');

    alert.close = async () => {
      await this.proxyCommand(`/element/${_appiumSupport.util.unwrapElement(closeButton)}/click`, 'POST');
    };
  } else {
    const firstButton = possibleAlertButtons[0];
    await assertButtonName(firstButton, 'cancel');
    const secondButton = possibleAlertButtons[1];
    await assertButtonName(secondButton, 'ok');

    alert.cancel = async () => {
      await this.proxyCommand(`/element/${_appiumSupport.util.unwrapElement(firstButton)}/click`, 'POST');
    };

    alert.ok = async () => {
      await this.proxyCommand(`/element/${_appiumSupport.util.unwrapElement(secondButton)}/click`, 'POST');
    };
  }

  alert.getText = async () => {
    const textViews = await this.findNativeElementOrElements('class name', 'XCUIElementTypeTextView', true, _appiumSupport.util.unwrapElement(alert));
    return await this.proxyCommand(`/element/${_appiumSupport.util.unwrapElement(_lodash.default.last(textViews))}/attribute/value`, 'GET');
  };

  alert.setText = async value => {
    const textViews = await this.findNativeElementOrElements('class name', 'XCUIElementTypeTextField', true, _appiumSupport.util.unwrapElement(alert));

    if (textViews.length === 0) {
      throw new Error('Tried to set text of an alert that was not a prompt');
    }

    await this.proxyCommand(`/element/${_appiumSupport.util.unwrapElement(textViews[0])}/value `, 'POST', {
      value
    });
  };

  return alert;
};

Object.assign(extensions, commands, helpers);
var _default = extensions;
exports.default = _default;require('source-map-support').install();
